### PR TITLE
New decoding method

### PR DESCRIPTION
**Pull request by knzo25**
## PR Type

- Improvement

## Description
The RKHS method presented several miss-classifications and did not work on the XX1 data due to the nature of the lidar. We implemented a new method based on the RKHS decoding but with normalization per bit area and instead of arbitrary templates we just use the standard bit template from apriltag. On the available data is does not present any problems so far.

## Review Procedure

Confirm that it works in all products

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Code follows [coding guidelines][coding-guidelines]
- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets
- [ ] Write [release notes][release-notes]
- [ ] (When added something to `.repos`) Check if proper access rights are set

## CI Checks

- **vcs import**: Required to pass before the merge.
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: src/lidartag_decode.cpp [hunk 1]
```diff
@@ -1747,6 +1777,38 @@ bool LidarTag::decodePayload(ClusterFamily_t & cluster)
     }
   }
 
+  // Data collection logic
```
Comment (aohsato) [inline src/lidartag_decode.cpp]: remove?

File: test/test_hamming_decoding.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  double tag_size = 0.8;
+  int decoded_id = -1;
+  int decoded_orientation = -1;
+
+  // data
+  int expected_id = 0;
+  int expected_orientation = 2;
+  std::string template_path = "/home/kenzolobos/workspace/pilot-auto.x2/src/vendor/lidartag/lib/aip_x2/templates";
+  std::string pointcloud_path = "/home/kenzolobos/workspace/pilot-auto.x2/decoding_database/2022-04-22_x2_glp_moving/camera0/0/2/1650606240 _17_37.pcd";
```
Comment (aohsato) [inline test/test_hamming_decoding.cpp]: change default path

**Review by aohsato — COMMENTED** [review]
LGTM about logic, but please apply clang-format as first?

**Comment by knzo25** [comment]
@aohsato 
Thanks for the review.
Should I apply clang-format to the whole repo?
It is not originally ours so I was worried/wondering whether to do it or not

File: src/lidartag_decode.cpp (hunk 1 above)
Comment (knzo25) [reply src/lidartag_decode.cpp]: done :ok_man:

File: test/test_hamming_decoding.cpp (hunk 2 above)
Comment (knzo25) [reply test/test_hamming_decoding.cpp]: done :ok_man:

**Comment by aohsato** [comment]
@knzo25 
> It is not originally ours so I was worried/wondering whether to do it or not

Ah, sorry you're right. It's not needed to apply format currently.

File: src/decoding/naive_hamming_decoding.cpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+      return std::get<0>(lhs) > std::get<0>(rhs);
+    });
+
+  double best_score = std::get<0>(templates_decoded_bits[0]);
+  double best_margin =
+    std::get<0>(templates_decoded_bits[0]) - std::get<0>(templates_decoded_bits[1]);
+  int best_id = std::get<1>(templates_decoded_bits[0]);
+  int best_orientation = std::get<2>(templates_decoded_bits[0]);
+
+  //
```
Comment (aohsato) [inline src/decoding/naive_hamming_decoding.cpp]: remove or add comment?

Comment (knzo25) [reply src/decoding/naive_hamming_decoding.cpp]: :open_mouth:

**Review by aohsato — APPROVED** [review]
LGTM, good job!

**Comment by knzo25** [comment]
:tada: :tada: :tada:

**Pull request merged**